home/web/client/inject.js

```diff
@@ -16,34 +15,78 @@ import web3extensions from 'dapps-react-components/src/util/web3.extensions';
 import Web3Provider from 'dapps-react-components/src/Web3Provider';
 import Root from './components/Root';
 import Interceptor from './utils/Interceptor';
+import InterceptorProvider from './providers/interceptor';
+import DomProvider from './providers/dom';
+import RpcProvider from './providers/rpc';
+import ExtensionProvider from './providers/extension';
 import readInjectOptions from './utils/readInjectOptions';
+import createStore from './store/configureStore';
+import middlewares from './middlewares';
+import { initApp } from './actions/app';
 
 const http = new Web3.providers.HttpProvider('/rpc/');
 const interceptor = new Interceptor(http);
-const web3 = new Web3(interceptor);
-const rawWeb3 = new Web3(http);
-web3extensions(rawWeb3)
-  .map(extension => rawWeb3._extend(extension));
+const exposedWeb3 = new Web3(interceptor);
+
+const internalWeb3 = new Web3(http);
+web3extensions(internalWeb3)
+  .map(extension => internalWeb3._extend(extension));
 
 // expose global web3
-global.web3 = web3;
+global.web3 = exposedWeb3;
 const options = readInjectOptions();
 
 // Render account chooser
 const el = document.createElement('div');
 document.querySelector('html').appendChild(el);
 
+const store = createStore(middlewares(internalWeb3), { options });
+
+// Needed for onTouchTap, for material ui
+// http://stackoverflow.com/a/34015469/988941
+injectTapEventPlugin();
+
 ReactDOM.render(
-  // wrapping id used to resest css, see inject.css
-  <div className={ stylesReset.reset }>
-    <Web3Provider web3={ rawWeb3 }>
-      <MuiThemeProvider muiTheme={ muiTheme }>
-        <Root
-          interceptor={ interceptor }
-          options={ options }
-        />
-      </MuiThemeProvider>
-    </Web3Provider>
-  </div>,
+  <Provider store={ store }>
+    <div className={ stylesReset.reset }>
+      <Web3Provider web3={ internalWeb3 }>
+        <MuiThemeProvider muiTheme={ muiTheme }>
+          <Root />
+        </MuiThemeProvider>
+      </Web3Provider>
+    </div>
+  </Provider>,
   el
 );
+
+const domProvider = new DomProvider(store);
+domProvider.init();
+
+const rpcProvider = new RpcProvider(store, internalWeb3);
+rpcProvider.init();
+
+const extensionProvider = new ExtensionProvider(store);
+extensionProvider.init();
+
+// IMPORTANT! Interceptor provider must be set in synchronous order after exposing web3
```

Comment:
But all stuff that this comment relates to will be close to each other. Also IMHO it makes more sense to first initialize store and middlewares and then render the state.